src/tools/perf/perftest_mad.c

```diff
@@ -0,0 +1,694 @@
+    int perftest_rte_class = PERFTEST_RTE_CLASS;
+    int oui                = IB_OPENIB_OUI;
+    int rmpp_version       = UMAD_RMPP_VERSION;
+    struct ibmad_port *port;
+
+    if ((ca == NULL) || (ca_port < 0)) {
+        ucs_error("MAD: missing CA or CA Port");
+        return NULL;
+    }
+
+    port = mad_rpc_open_port(ca, ca_port, mgmt_classes, mgmt_classes_size);
+    if (port == 0) {
+        ucs_error("MAD: failed to open '%s:%d'", ca, ca_port);
+        return NULL;
+    }
+
+    if (mad_register_server_via(perftest_rte_class, rmpp_version, NULL, oui,
+                                port) < 0) {
+        ucs_error("MAD: cannot serve perftest RTE class 0x%02x on '%s:%d'",
+                  perftest_rte_class, ca, ca_port);
+        goto fail;
+    }
+    return port;
+fail:
+    mad_rpc_close_port(port);
+    return NULL;
+}
+
+void perftest_mad_close(struct ibmad_port *port)
+{
+    if (port != 0) {
+        mad_rpc_close_port(port);
+    }
+}
+
+static ucs_status_t perftest_mad_sm_query(const char *ca, int ca_port,
+                                          const struct ibmad_port *mad_port,
+                                          uint64_t guid, ib_portid_t *dst_port)
+{
+    uint8_t buf[IB_SA_DATA_SIZE] = {};
+    umad_port_t port             = {};
+    __be64 prefix;
+    ibmad_gid_t selfgid;
+    uint64_t port_guid;
+    uint64_t gid_prefix;
+    int ret;
+    ib_portid_t sm_id; /* SM: the GUID to LID resolver */
+
+    ret = umad_get_port(ca, ca_port, &port);
+    if (ret < 0) {
+        ucs_error("MAD: could not get SM LID");
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    memset(&sm_id, 0, sizeof(sm_id));
+    sm_id.lid = port.sm_lid;
+    sm_id.sl  = port.sm_sl;
+
+    memset(selfgid, 0, sizeof(selfgid)); /* uint8_t[] */
+    gid_prefix = be64toh(port.gid_prefix);
+    port_guid  = be64toh(port.port_guid);
+
+    umad_release_port(&port);
+
+    mad_encode_field(selfgid, IB_GID_PREFIX_F, &gid_prefix);
+    mad_encode_field(selfgid, IB_GID_GUID_F, &port_guid);
+
+    memcpy(&prefix, selfgid, sizeof(prefix));
+    mad_set_field64(dst_port->gid, 0, IB_GID_PREFIX_F,
+                    prefix ? be64toh(prefix) : IB_DEFAULT_SUBN_PREFIX);
+    mad_set_field64(dst_port->gid, 0, IB_GID_GUID_F, guid);
+
+    dst_port->lid = ib_path_query_via(mad_port, selfgid, dst_port->gid, &sm_id,
+                                      buf);
+    if (dst_port->lid < 0) {
+        ucs_error("MAD: GUID query failed");
+        return UCS_ERR_UNREACHABLE;
+    }
+    mad_decode_field(buf, IB_SA_PR_SL_F, &dst_port->sl);
+    return UCS_OK;
+}
+
+static ucs_status_t perftest_mad_get_portid(const char *ca, int ca_port,
+                                            const char *addr,
+                                            const struct ibmad_port *mad_port,
+                                            ib_portid_t *dst_port)
+{
+    static const char guid_str[] = "guid:";
+    static const char lid_str[]  = "lid:";
```

Comment:
```
$ ibstat mlx5_0 | grep -E '(Base lid|Port GUID)'
                Base lid: 46
                Port GUID: 0x98039b030096b174
```
Supported format:
```
ucx_perftest -K mlx5_0:1 -t ucp_put -e lid:0x49
ucx_perftest -K mlx5_0:1 -t ucp_put -e lid:73
ucx_perftest -K mlx5_0:1 -t ucp_put -e guid:0x98039b030096b174
```
Documented on format failure like:
```
perftest_mad.c:529  UCX  ERROR MAD: invalid dst address, use 'guid:' or 'lid:' prefix
```

Reply:
Adding `:` on the command line for the guid would be cumbersome instead of doing direct copy/paste from ibstat output.

Without the prefix, since lid and guid currently share the same format, to differentiate, we would need to rely on the size which seems to be weak, since lid can allow ~48k nodes. On the documentation part we could also add to the help message. What do you think?